src/main/java/org/testng/IModule.java

```diff
@@ -0,0 +1,9 @@
+package org.testng;
+
+import com.google.inject.Module;
+
+/**
+ * This is a marker interface for a Guice module to be instantiated with {@link java.util.ServiceLoader}.
+ */
+public interface IModule extends Module {
```

Comment:
I'm not feeling good to have IModule extending Guice Module: we can easily imagine having module from app code but we don't want to have a dependency to a test-lib in the app code. 
Instead, what do you think to have an IModelLoader with a method `List<Module> getModules()`?

Reply:
Actually, the issue emerged with TestNG 6.11. I had to use com.guice.inject.Injector in my class. Because Guice creates JIT bindings in the parent injector and TestNG prior to version 7.0.1 creates a parent injector aside from the com.google.inject.Guice defined modules injector, my class had the former one injected. Since https://github.com/cbeust/testng/issues/2017 has been fixed this case should not be an issue anymore. 
If I could explicitly bind my class, I could fix this issue transparently for a user.
I believe it would be useful in other cases as well, eg. if JIT binding is not sufficient because of no @Inject annotated constructor.
I got your point, this interface may condone a bad application design, but the same way a user may implement IModuleLoader in the application Module class. Additionally, once a user is to implement IModule that will be used solely in tests, he would have to implement additional IModuleLoader. It seems unnecessary.
The proposed getModules() method returns a list of modules. In the case of IModule, a user may implement a single module that installs (com.google.inject.Module#install) other modules.

Reply:
> the same way a user may implement IModuleLoader in the application Module class

No, it shouldn't be the case because my idea is more:
```
src/.../XxxModule.java
test/.../AppModuleLoader.java
```
where AppModuleLoader will provide XxxModule(and others)

```java
public List<Class<? extends Module>> getModules() {
   return List.of(XxxModule.class);
}
```

With your solution, if you want separate concerns, I only see a `TestXxxModule extends XxxModule and implements IModule`. I think it will work but won't be easy to explain.

I prefer a design where the test execution is configured instead of using language extensions.

@krmahadevan thought?

Reply:
TestXxxModules may just invoke install(new XxxModule()) instead inheriting XxxModule.

Reply:
@krmahadevan any update? One more case this feature would be beneficial is when IInjectorFactory creates an injector with a parent injector set.

Reply:
@bartoszpop I haven't found time to take a closer look at this. Been swamped with things at work front. Will take a look at the earliest.

Reply:
@bartoszpop - I had a closer look at this changeset. Is there any specific reason behind us creating `org.testng.IModule` as a marker interface that just extends `com.google.inject.Module` ?

Why couldn't we have just resorted to using `com.google.inject.Module` for loading via SPI ?

The other question I had was, that usually things that get loaded via SPI are expected to happen once per JVM, but here, it seems to be happening multiple times. 

Can we not have this changed such that its lazy loaded and done only once per JVM ?

Reply:
The reason is that com.google.inject.Module is not a TestNG interface and Guice or some other library may also load modules via SPI and these modules may clash. It is better to have a TestNG specific SPI configuration file (META-INF/services/org.testng.IModule).

> Can we not have this changed such that its lazy loaded and done only once per JVM ?
Done

Reply:
@bartoszpop 

> The reason is that com.google.inject.Module is not a TestNG interface and Guice or some other library may also load modules via SPI and these modules may clash. 

Fair enough. In that case, I think we should just prefer defining an interface that returns back module

```java
public interface IModule {
  Module getModule();
}
```